### PR TITLE
Add useful errors when model is not configured correctly

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/1171 and https://github.com/PyTorchLightning/pytorch-lightning/issues/508

This PR adds checks, that secure that the users model is correctly configured before training. This includes:
* Checking that `training_step` has been overridden
* Checking that `training_dataloader` has been overridden
* Warning if `configure_optimizers` has not been overridden, will tell user that program is using default optimizer (adam with lr=0.0001)
* Error if a `validation_dataloader` is overridden but no `validation_step` is defined (and vise verse)
* Error if a `test_dataloader` is overriden but no `test_step` is defined (and vise verse)
* Warning if `validation_dataloader, validation_step` is overridden but no `validation_epoch_end`
* Warning if `test_dataloader, test_step` is overridden but no `test_epoch_end`

The most fundamental change is the requirement of `validation_step` and `test_step` when there respective dataloaders are defined. This will probably not be backward compatible with some users code.

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -886,6 +877,66 @@ def test(self, model: Optional[LightningModule] = None):
...
+
+        if not self.is_overriden('train_dataloader', model):
+            m = ('No train_dataloader() defined. Lightning expects as minimum '
+                 'a training_step() and training_dataloader() to be defined.')
+            raise MisconfigurationException(m)
+
+        if not self.is_overriden('configure_optimizers', model):
+            m = ('configure_optimizers() method not defined by user, will default '
+                 'to Adam optimizer with learning rate set to 0.0001.')
+            warnings.warn(m)
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: maybe `RuntimeWarning`?

**Comment by Borda** [comment]
@neggert I guess you already were solving the DDP pickle issue, right?
`tests/test_gpu_models.py::test_ddp_all_dataloaders_passed_to_fit`
tells:
```
obj = <SpawnProcess(SpawnProcess-9, initial)> | 811s
-- | --
920 | file = <_io.BytesIO object at 0x7fa73c5d26d0>, protocol = None | 811s
921 |   | 811s
922 | def dump(obj, file, protocol=None): | 811s
923 | '''Replacement for pickle.dump() using ForkingPickler.''' | 811s
924 | >       ForkingPickler(file, protocol).dump(obj) | 811s
925 | E       TypeError: can't pickle code objects
```
@SkafteNicki may you pls resolve conflist in chnagelog?

**Comment by SkafteNicki** [comment]
@Borda I update code with your recommendations and have solved the conflict. Only the pickle issues seems to be blocking a merge.

**Comment by SkafteNicki** [comment]
@jeremyjordan I have added the test that you requested

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support for hierarchical `dict` ([#1152](https://github.com/PyTorchLightning/pytorch-lightning/pull/1152))
 - Added `TrainsLogger` class ([#1122](https://github.com/PyTorchLightning/pytorch-lightning/pull/1122))
 - Added type hints to `pytorch_lightning.core` ([#946](https://github.com/PyTorchLightning/pytorch-lightning/pull/946))
+- Added support for IterableDataset in validation and testing ([#1104](https://github.com/PyTorchLightning/pytorch-lightning/pull/1104))
```
Comment (ethanwharris) [inline CHANGELOG.md]: Looks like these 3 lines got accidentally duplicated

**Comment by jeremyjordan** [comment]
some related work going on in #1279

**Comment by SkafteNicki** [comment]
@ethanwharris I can see that you are in charge of #1279, and I can see that we both have included warnings regarding `configure_optimizers`. We should probably sync, so the user is not warned twice with different messages.

**Comment by ethanwharris** [comment]
@SkafteNicki Yeah, I think the best thing would be to let this merge first and then I'll update #1279 to work on top (and make sure there are no problems with warnings etc.) I'll switch it to a draft for now :)

File: pl_examples/basic_examples/lightning_module_template.py [hunk 3]
```diff
@@ -241,6 +241,38 @@ def test_dataloader(self):
...
+        """
+        results = self.validation_step_end(outputs)
+
+        # Rename output
+        tqdm_dict = results['progress_bar']
+        tqdm_dict['test_loss'] = tqdm_dict.pop('val_loss')
+        tqdm_dict['test_acc'] = tqdm_dict.pop('val_acc')
+
+        results['progress_bar'] = tqdm_dict
+        results['log'] = tqdm_dict
```
Comment (Borda) [inline pl_examples/basic_examples/lightning_module_template.py]:
```suggestion
        # rename some keys
        results['progress_bar'].update({
            'test_loss': results['progress_bar'].pop('val_loss'),
            'test_acc': tqdm_dict.pop('val_acc'),
        })

        results['log'] =  results['progress_bar']
```

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -961,6 +952,62 @@ def test(self, model: Optional[LightningModule] = None):
...
+
+        Args:
+            model: The model to test.
+
+        """
+        # Check training_step, train_dataloader, configure_optimizer methods
+        if not self.is_overriden('training_step', model):
+            raise MisconfigurationException('No training_step() method defined.'
+                                            ' Lightning expects as minimum a `training_step()` and'
+                                            ' `training_dataloader()` to be defined.')
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]:
```suggestion
            raise MisconfigurationException(
                'No `training_step()` method defined. Lightning expects as minimum a `training_step()` and'
                ' `training_dataloader()` to be defined.')
```

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -961,6 +952,62 @@ def test(self, model: Optional[LightningModule] = None):
...
+
+        if not self.is_overriden('train_dataloader', model):
+            raise MisconfigurationException('No train_dataloader() defined.'
+                                            ' Lightning expects as minimum a `training_step()` and'
+                                            ' `training_dataloader()` to be defined.')
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]:
```suggestion
            raise MisconfigurationException(
                'No `train_dataloader()` defined. Lightning expects as minimum a `training_step()` and'
                ' `training_dataloader()` to be defined.')
```

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -624,3 +628,123 @@ def test_epoch_end(self, outputs):
 
     model = LightningTestModel(hparams)
     Trainer().test(model)
+
+
+def test_error_on_no_train_step(tmpdir):
```
Comment (Borda) [inline tests/trainer/test_trainer.py]: let's move it to new module `test_checks`

**Comment by awaelchli** [comment]
Awesome!

**Comment by Borda** [comment]
@SkafteNicki nice job, could you pls resolve conflicts...

**Comment by SkafteNicki** [comment]
@Borda conflicts are solved now, some (unrelated) tests are still failing...

**Comment by Borda** [comment]
@SkafteNicki I have restarted the GitHub, but the GPU is failing because of
```
def dump(obj, file, protocol=None): | 801s
-- | --
1709 | '''Replacement for pickle.dump() using ForkingPickler.''' | 801s
1710 | >       ForkingPickler(file, protocol).dump(obj) | 801s
1711 | E       TypeError: can't pickle code objects
```
I thought we were fixing it... @neggert?

**Comment by Borda** [comment]
could we merge this one and then Will's as followup?

**Comment by SkafteNicki** [comment]
As the checks implemented in this PR is only called within `Trainer.fit()` I think that they should indeed throw an error instead of a warning. This means that the user can freely use LightningModules as any other Pytorch modules until they try to make use of lightnings trainer functionality, in which we enforce some structure. 

If you want to merge #1317 into this, fine by me. I will update `configure_optimizers` error/warning afterwards.

**Comment by williamFalcon** [comment]
@SkafteNicki awesome!

**Comment by Borda** [comment]
@williamFalcon @SkafteNicki there is still a bug and now it is in master
```
def dump(obj, file, protocol=None): | 1272s
-- | --
1712 | '''Replacement for pickle.dump() using ForkingPickler.''' | 1272s
1713 | >       ForkingPickler(file, protocol).dump(obj) | 1272s
1714 | E       TypeError: can't pickle code objects
```
http://35.192.60.23/PyTorchLightning/pytorch-lightning/954/1/2

**Comment by williamFalcon** [comment]
probably that __code__ call

**Comment by neggert** [comment]
Yeah `test_ddp_all_dataloaders_passed_to_fit` is failing. I'd imagine it's this line: 

```python
self.__code__ = self.__call__.__code__
```

Seems that code object isn't pickleable. We need everything to be pickleable for DDP to work.

**Comment by Borda** [comment]
so converting it to a string would help?

**Comment by Borda** [comment]
@bmartinn with removing the line above we are getting Trains issue:
```
Traceback (most recent call last):
  File "/home/software/miniconda3/envs/pl10/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/home/software/miniconda3/envs/pl10/lib/python3.7/site-packages/trains/backend_interface/logger.py", line 197, in run
    while not self._exit_event.wait(period or 1.0):
  File "/home/software/miniconda3/envs/pl10/lib/python3.7/site-packages/trains/logger.py", line 522, in flush
    return self._task.flush()
  File "/home/software/miniconda3/envs/pl10/lib/python3.7/site-packages/trains/task.py", line 751, in flush
    LoggerRoot.flush()
  File "/home/software/miniconda3/envs/pl10/lib/python3.7/site-packages/trains/debugging/log.py", line 92, in flush
    h.flush()
  File "/home/software/miniconda3/envs/pl10/lib/python3.7/logging/__init__.py", line 1009, in flush
    self.stream.flush()
ValueError: I/O operation on closed file.
```

**Comment by tbachlechner** [comment]
It appears that `training_dataloader` does not exist, I think it should read `train_dataloader`? Another question is naming convention consistency with `training_step`.

**Pull request merged**